pkg/octtrpc/interceptor.go

```diff
@@ -98,20 +111,33 @@ func ServerInterceptor(opts ...Option) ttrpc.UnaryServerInterceptor {
 	for _, opt := range opts {
 		opt(&o)
 	}
-	return func(ctx context.Context, unmarshal ttrpc.Unmarshaler, info *ttrpc.UnaryServerInfo, method ttrpc.Method) (_ interface{}, err error) {
+
+	return func(ctx context.Context, unmarshal ttrpc.Unmarshaler, info *ttrpc.UnaryServerInfo, method ttrpc.Method) (resp interface{}, err error) {
 		name := convertMethodName(info.FullMethod)
 
 		var span *trace.Span
 		opts := []trace.StartOption{trace.WithSampler(o.sampler), oc.WithServerSpanKind}
-		parent, ok := getParentSpanFromContext(ctx)
-		if ok {
+		if parent, ok := getParentSpanFromContext(ctx); ok {
 			ctx, span = oc.StartSpanWithRemoteParent(ctx, name, parent, opts...)
 		} else {
 			ctx, span = oc.StartSpan(ctx, name, opts...)
 		}
 		defer span.End()
-		defer setSpanStatus(span, err)
+		defer func() {
+			if err == nil {
+				span.AddAttributes(trace.StringAttribute("response", log.Format(ctx, resp)))
+			}
+			setSpanStatus(span, err)
+		}()
+		if len(o.attrs) > 0 {
+			span.AddAttributes(o.attrs...)
+		}
 
-		return method(ctx, unmarshal)
+		return method(ctx, func(req interface{}) (err error) {
+			if err = unmarshal(req); err == nil {
+				span.AddAttributes(trace.StringAttribute("request", log.Format(ctx, req)))
+			}
+			return err
+		})
```

Comment:
I'm concerned about always logging the request/response:
- These could be quite large (compared to a normal span) in many cases, which would bloat logging data or lead to worse performance.
- These could contain sensitive information that should not be logged.

Reply:
I'll validate/update the scrubbing for the requests that contain spans, but we already log most components of the pieces of the requests/responses in the span, but we manually add the fields:
https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/cmd/containerd-shim-runhcs-v1/service.go#L126-L150

The only exception is Exec process, which carries the OCI spec, and that we log latter on regardless.
Also, we log all requests/responses over the bridge (and log them again on the Linux GCS side):

https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/internal/gcs/bridge.go#L399-L402
https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/internal/guest/bridge/bridge.go#L322

https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/internal/gcs/bridge.go#L298-L301
https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/internal/guest/bridge/bridge.go#L386

Performance wise, the formatting uses `google.golang.org/protobuf/encoding/protojson`, which should be quite performant:
https://github.com/microsoft/hcsshim/blob/fa3b77dcc71664d0594e3d0459c17e439b005fec/internal/log/format.go#L71-L79